Using team namespaces: Kubeconfig Step Thru

mplsgrant opened the pull request:
I've created this as an extension on #598. It includes 
- some light admin documentation 
- an update to user permissions
- an updated `deploy` function to create tanks in a specific namespace and "--to-all-users"
- an updated `down` function that asks before nuking a whole wargame (and a "--force" to not bother asking)
- e2e test 
- handling of namespace throughout many other functions: status, logs, etc
- connecting nodes across namespaces using dns: tank-0003.wargames-blue-team.svc.cluster.local
- fix to continue checking the init container if at first it can't be found
- solves #618

josibake committed:
refactor: namespaces.yaml, namespace-defaults.yaml

namespaces.yaml is meant for describing the overall structure of what you want
with specific overrides for specific users as needed. the "default" roles should
be defined in namespace-defaults.yaml so that they are automatically applied by default
for each user in each namespace.

at a lower level, defaults that should be applied by default for *any* namespaces deployment
should be defined in values.yaml.

namespace-defaults.yaml is meant to override values.yaml in the event
for a particular namespaces deployment the admin wants to create
tailor made roles and permisssions. otherwise, this can stay empty
and whatever is in values.yaml will be applied.

update example prefix to wargames, to illustrate this is not relying on
a default namespace of warnet. this probably needs some more thought,
but I think its best to address how to pipe through the name in a
followup rather than slow this PR down.

josibake committed:
deploy: allow a namespace parameter in `deploy`

Allowing a namespace parameter in deploy means that we can easily launch tanks into a given
namespace, and we can safely ignore this parameter in other functions because it defaults to None
which the code will handle gracefully.

mplsgrant committed:
admin.md: add admin documentation

mplsgrant committed:
namespaces: remove the dir from the `namespaces`

I don't think the dir argument is necessary because we are just querying the cluster.

mplsgrant committed:
charts: allow users access to services

We will need to allow users access to services for launching scenarios.

mplsgrant committed:
charts: add namespace to sa permissions

mplsgrant committed:
status: add namespace to `status`

mplsgrant committed:
constants: make the default namespace "default"

The general agreement we came to was to use "default" as the default namespace. I took a moment to
search the code base for instances of hard-coded "warnet" namespaces, and I could only really find
one. It was in the commander file -- specifically in the network flag area. I updated this in another commit.

mplsgrant committed:
k8s: fix static_client type

mplsgrant committed:
constants: add `wargames` prefix

Until we start using labels in earnest for labeling and querying namespaces, this prefix will have
to suffice for our labeling and querying needs.

mplsgrant committed:
k8s: update getting namespace logic

mplsgrant committed:
control: use namespace log in `down`

mplsgrant commented:
@josibake Made my changes in this Stepthru pull. The prefix logic in #598 wasn't working like I wanted.
I plan to look into this again Fri afternoon.

josibake commented:
Nice, took a brief look and this looks good! I think the prefix approach is fine, and then we can revisit in a follow up a more comprehensive approach based on labels/annotations, which would remove the need for any sort of naming convention in our namespaces.

mplsgrant committed:
constants: add k8s internal namespaces

We want to know the namespaces used internally by k8s so that we can filter them out when querying
namespaces.

We can also likely replace this logic after we create a labeling strategy for our namespaces.

mplsgrant committed:
k8s: ignore internal namespaces

mplsgrant committed:
control & test base: prevent hasty `down`s

This does a check to ensure users don't nuke the network by entering `warnet down`. It will prompt
the users if there are multiple namespaces affected. Also adds a `--force` option to bypass this.

mplsgrant committed:
namespaces: flesh out wargames prefix

mplsgrant committed:
testing: add e2e namespace/admin test

mplsgrant committed:
workflow: add test to git workflow

mplsgrant committed:
control: clean up `down` command

mplsgrant committed:
k8s & control: query all namespaces

This expands `get_pods` view to query all namespaces.

It also flattens out list[V1PodList] into just list[V1Pod]. This affected two downstream functions:
`down` and `_logs`. I updated them accordingly.

mplsgrant committed:
test base: add namespace to get_pod_exit_status

mplsgrant committed:
k8s: add optional ns to get_pod_exit_stats

mplsgrant committed:
k8s: add optional ns to delete_pod

mplsgrant committed:
k8s: add optional ns to get_edges

mplsgrant committed:
k8s: add optional ns to snapshot_bitcoin_datadir

mplsgrant committed:
k8s: add optional ns to wait_for_init

mplsgrant committed:
k8s: add optional ns to pod_log

mplsgrant committed:
k8s: add optional ns to wait_for_pod

mplsgrant committed:
k8s: add optional ns to write_file_to_container

mplsgrant committed:
network: add namespace to `network` `_connected`

mplsgrant committed:
control: add ns to `logs`

mplsgrant committed:
control: add namespace to `run`

mplsgrant committed:
control: add imports

mplsgrant committed:
bitcoin: add ns to `rpc`

mplsgrant committed:
bitcoin: add ns to `debug_log` and `grep_logs`

mplsgrant committed:
bitcoin: add ns to `messages` and `get_messages`

josibake commented:
Did a walk through of the commits, looks good. I think it would be helpful to squash the commits into logical groups, e.g., one commit for adding the ns option to the cli arguments, one commit for all of the namespace prefix logic, etc. I found it a bit confusing to step through each commit where later commits were re-editing lines of code that had already been touched.

Alternatively, people can just review the files, but given the scope of the PR I think logical commits would be helpful.

mplsgrant committed:
service_accounts: add sa func to admin section

mplsgrant committed:
k8s: add ns, sa, and config helper funcs

mplsgrant committed:
admin: add create_kubeconfigs func

mplsgrant committed:
k8s: add namespace to `wait_for_init`

mplsgrant commented:
@josibake Updated the commit structure as you suggested.

mplsgrant committed:
k8s: add `continue` to `wait_for_init`

When this function encounters a matching pod, it attempts to retrieve that pod's
init_container_statuses. The problem seems to be that this metadata is not available (perhaps
because the pod is not read?). This means that the metadata is set to None which is not iterable and
therefore causes a crash.

This fix allows the program to proceed to the next element in the for loop stream if the init_container_statuses metadata
is not ready (aka is None).

mplsgrant committed:
control: ignore logging namespaces

mplsgrant committed:
deploy: enable deploying to all user namespaces

mplsgrant commented:
@josibake This PR is looking good I think. Added a flag to `deploy` so that we can easily deploy the network to multiply namespaces at once.

mplsgrant commented:
rebased on main

pinheadmz commented on src/warnet/admin.py:
```diff
@@ -33,3 +35,93 @@ def init():
+    namespace: {namespace}
+    user: {sa}
+current-context: {sa}-{namespace}
+"""
```
might be cleaner to implement as a dict and then use `yaml.dump()`

mplsgrant replied:
Fixed here: https://github.com/bitcoin-dev-project/warnet/pull/616/commits/f50dcf82335ce8c24a23b0e8a2775b7df5686537 "admin: make kubeconfig a dict"

pinheadmz reviewed (commented):
ok this is a lot, i breezed through each commit and added one comment. going to focus on test and docs next

pinheadmz commented on src/warnet/bitcoin.py:
```diff
@@ -23,23 +24,24 @@ def bitcoin():
-    namespace = get_default_namespace()
-
+    if not namespace:
+        namespace = get_default_namespace()
```
i wonder if the repetition of this can be smoothed out with a function decorator like `@needs_namespace` that can be applied to all relevant commands

mplsgrant replied:
Fixed here: https://github.com/bitcoin-dev-project/warnet/pull/616/commits/2a906e28b5ed17efe74a75976c4781507b501af4 "DRY out the namespace check"

pinheadmz commented on src/warnet/admin.py:
```diff
@@ -33,3 +37,93 @@ def init():
+
+    # Get all namespaces that start with prefix
+    # This assumes when deploying multiple namespacs for the purpose of team games, all namespaces start with a prefix,
+    # e.g., tabconf-wargames-*. Currently, this is a bit brittle, but we can improve on this in the future
```
I'm still pretty worried about the tech debt from this mechanism since it is so ubiquitous in this PR. Hard-coding the k8s namespaces and logging namesapce as well... and forcing the user to use a naming convention... all seems to go against our design principles!

josibake replied:
Agree that the prefixes is an anti pattern. I was looking at using namespace labels so they we could do things like `get_namespaces_by_type(WARNET_TANKS)` / `get_namespaces_by_type(LOGGING)` / `get_namespaces_by_type(ALL)`, etc but that felt like it would take more work to get right. If the tradeoff is we have prefixed namespaces and can use namespaces for Tabconf, vs we don't have namespaces for Tabconf at all, I'm more inclined to proceed as is and have a backlog to refactor to using labels (happy to own following up on that).

pinheadmz commented:
test fails locally bc im using docker desktop:

```

Network connected
2024-10-03 08:48:27 | INFO    | test     | Authenticating and becoming bob...
2024-10-03 08:48:27 | INFO    | test     | Stopping network
2024-10-03 08:48:27 | DEBUG   | test     | Executing warnet command: down --force
2024-10-03 08:48:28 | DEBUG   | test     | Waiting for predicate with timeout 60s and interval 1s
2024-10-03 08:48:29 | INFO    | test     | Waiting for all tanks to reach 'stopped': {'total': 1, 'running': 1}
Traceback (most recent call last):
  File "/Users/matthewzipkin/Desktop/work/warnet/test/namespace_admin_test.py", line 124, in <module>
    test.run_test()
  File "/Users/matthewzipkin/Desktop/work/warnet/test/namespace_admin_test.py", line 32, in run_test
    self.authenticate_and_become_bob()
  File "/Users/matthewzipkin/Desktop/work/warnet/test/namespace_admin_test.py", line 69, in authenticate_and_become_bob
    assert get_kubeconfig_value("{.current-context}") == "minikube"
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError

```

josibake commented on src/warnet/bitcoin.py:
```diff
@@ -126,16 +139,29 @@ def grep_logs(pattern: str, show_k8s_timestamps: bool, no_sort: bool):
 @click.option("--chain", default="regtest", show_default=True)
-def messages(tank_a: str, tank_b: str, chain: str):
+@click.option("--namespace_a", default=None, show_default=True)
+@click.option("--namespace_b", default=None, show_default=True)
```
Is this necessary if we instead just passed the fully tank hostname, e.g., `tank.namespace.svc`?

mplsgrant replied:
I would like to be able to specify the bare tank name and namespace. Maybe we could also parse the tank name to determine if it is a dns name, and act accordingly. 

I would also like to give `messages` the `logs` treatment.

josibake commented on src/warnet/bitcoin.py:
```diff
@@ -160,23 +186,25 @@ def messages(tank_a: str, tank_b: str, chain: str):
 
 
-def get_messages(tank_a: str, tank_b: str, chain: str):
+def get_messages(tank_a: str, tank_b: str, chain: str, namespace_a: str, namespace_b: str):
```
same comment regarding fully tank hostname. Might be slightly annoying to have to type out the full name if you're only operating on one namespace (majority use case), but the plus side is simpler code?

josibake commented on src/warnet/control.py:
```diff
@@ -284,35 +335,43 @@ def filter(path):
+            pod_list = [
+                f"{item.metadata.name}: {item.metadata.namespace}"
+                for item in pods
+                if item.metadata.namespace not in [LOGGING_NAMESPACE, INGRESS_NAMESPACE]
```
would be nice if this logic was inside some sort of `get_namespaces` function, instead of the RPC command itself. It seems like we will eventually want to get rid of any sort of hardcoding/namespace conventions, so hiding all of the hardcoded logic inside functions will make this _much_ easier to refactor in the future.

mplsgrant replied:
Fixed. Simplified the logic using labels here: https://github.com/bitcoin-dev-project/warnet/pull/616/commits/f7e43c47f83d763660cd4ac550e1f42ed59c799c "constants: use labels to select pods for `log`"

josibake commented on src/warnet/deploy.py:
```diff
@@ -42,13 +49,31 @@ def validate_directory(ctx, param, value):
 @click.option("--debug", is_flag=True)
-def deploy(directory, debug):
+@click.option("--namespace", type=str, help="Specify a namespace in which to deploy the network")
+@click.option("--to-all-users", is_flag=True, help="Deploy network to all user namespaces")
```
I'm not crazy about this option.. it feels like we are building a use case specific convenience flag into the CLI. What's the advantage of having this vs just writing a simple bash script that deploys the network to each namespace by calling `warnet deploy /path/to/network --namespace X; warnet deploy path/to/network --namespace Y` ?

josibake replied:
I also quite like that `warnet deploy` is somewhat generic in that it can deploy namespaces / networks. but now this flag is specific to only deploying networks.

mplsgrant replied:
What if we had a `deploy_to_users()` function in the `admin` section? I would like to avoid using bash scripts.

pinheadmz replied:
i like `--to-all-users` personally

josibake commented on src/warnet/k8s.py:
```diff
@@ -359,3 +388,43 @@ def write_file_to_container(pod_name, container_name, dst_path, data):
+            return []
+
+
+def get_namespaces_by_prefix(prefix: str) -> list[V1Namespace]:
```
if the plan is to eventually move to labels, we could call this function `get_namespaces_by_type` or something similar, and then have this function be a black box. Right now its using prefix, but we could refactor just this function in the future to actually use labels/annotations.

mplsgrant replied:
Refactored here: https://github.com/bitcoin-dev-project/warnet/pull/616/commits/049e2dfebfe3e5307c3688aaa176226c7072826a "rename to `get_namespaces_by_type`"

josibake commented on src/warnet/service_accounts.py:
```diff
@@ -0,0 +1,38 @@
+    """Service account commands"""
+
+
+@service_accounts.command()
```
What's the use case for having this? Is this just to avoid running a kubectl command?

mplsgrant replied:
Yes. Also, if we move towards an `-o json` flag then this becomes more powerful and helpful within the code base.

josibake replied:
Hm, I'm not seeing how this is helpful. From what I can tell, this command would only be ran by an admin, and it's unclear what they would do with the output of this command , without needing to run more kubectl commands. If the argument is we are going to keep expanding our codebase to do k8s administrative work, I'd say this seems like we are re-writing kubectl in warnet and only for an admin user, which doesn't seem like a good idea to me.

mplsgrant replied:
Sounds good. Removed it here: https://github.com/bitcoin-dev-project/warnet/pull/616/commits/d29b30e03259b2bc286f63b8b349232311fcd4c2 and https://github.com/bitcoin-dev-project/warnet/pull/616/commits/78fa9050c51395e3fa6b25ca792f3b5623e656cc

pinheadmz commented on test/namespace_admin_test.py:
```diff
@@ -0,0 +1,124 @@
+class NamespaceAdminTest(TestBase):
+    def __init__(self):
+        super().__init__()
+        self.namespace_dir = (
```
the test in https://github.com/bitcoin-dev-project/warnet/pull/622 would be a good model for this one as well. after running the test locally I end up with a `kubeconfigs/` directory in the repo root. The test should operate in /tmp

mplsgrant replied:
Thank! Fixed in https://github.com/bitcoin-dev-project/warnet/pull/616/commits/64d1b40bd4f9bd345cb481dcdaa875e08c1b00c0 "tesing: use a temporary directory"

josibake commented on src/warnet/admin.py:
```diff
@@ -33,3 +37,93 @@ def init():
+    help="Duration of the token in seconds (default: 48 hours)",
+)
+def create_kubeconfigs(kubeconfig_dir, token_duration):
+    """Create kubeconfig files for all ServiceAccounts in warnet team namespaces starting with <prefix>."""
```
help is outdated (mentions passing a prefix argument)

mplsgrant replied:
Thank you. Fix here https://github.com/bitcoin-dev-project/warnet/pull/616/commits/9b1d070e4bdc58e7bef8821c33bc6c4572a84e88

josibake commented:
Some testing notes:

1. deployed namespaces
2. created kube-configs
3. authenticated as Alice (`warnet auth <kubeconfig>`
4. ran `warnet run scenario/mining_std.py --allnodes`

error:

```
warnet run scenarios/miner_std.py --allnodes
Including: __init__.py
Including: commander.py
Including: miner_std.py
Including: test_framework
Including: test_framework/__init__.py
Including: test_framework/address.py
Including: test_framework/authproxy.py
Including: test_framework/bdb.py
Including: test_framework/blockfilter.py
Including: test_framework/blocktools.py
Including: test_framework/coverage.py
Including: test_framework/descriptors.py
Including: test_framework/ellswift.py
Including: test_framework/key.py
Including: test_framework/messages.py
Including: test_framework/muhash.py
Including: test_framework/netutil.py
Including: test_framework/p2p.py
Including: test_framework/psbt.py
Including: test_framework/ripemd160.py
Including: test_framework/script.py
Including: test_framework/script_util.py
Including: test_framework/secp256k1.py
Including: test_framework/segwit_addr.py
Including: test_framework/siphash.py
Including: test_framework/socks5.py
Including: test_framework/test_framework.py
Including: test_framework/test_node.py
Including: test_framework/test_shell.py
Including: test_framework/util.py
Including: test_framework/wallet.py
Including: test_framework/wallet_util.py
Failed to deploy scenario commander: miner_std
Error: Error: create: failed to create: secrets is forbidden: User "system:serviceaccount:wargames-red-team:alice" cannot create resource "secrets" in API group "" in the namespace "wargames-red-team"
```

will need to dig into the commander image more, but kinda weird that the commander image needs to create "secrets"? Easy fix is to just change the default permissions for the role to allow users to create secrets in their namespaces, but would also like to understand why this permission is needed in the first place.

josibake commented:
Also, when running `warnet down`, it doesn't actually delete any resources (expected, per the permissions we gave Alice), but it tells the user that it did do the teardown successfully (lies!):

```
warnet down
Preparing to bring down the running Warnet...
The listed namespaces will be affected:
  wargames-red-team
[?] Do you want to bring down the running Warnet? (y/N): y

Bringing down the warnet...

Initiated uninstall for: tank-0001 in namespace wargames-red-team
Initiated uninstall for: tank-0003 in namespace wargames-red-team
Initiated uninstall for: tank-0002 in namespace wargames-red-team
Initiated deletion of pod: tank-0002 in namespace wargames-red-team
Initiated deletion of pod: tank-0001 in namespace wargames-red-team
Initiated deletion of pod: tank-0003 in namespace wargames-red-team
Teardown process initiated for all components.
Note: Some processes may continue in the background.
Warnet teardown process completed.
```

mplsgrant commented:
rebased on main

mplsgrant commented:
> Also, when running `warnet down`, it doesn't actually delete any resources (expected, per the permissions we gave Alice), but it tells the user that it did do the teardown successfully (lies!):

@josibake I supposed they are lies by omission. After all, it did initiate the deletion. The teardown process did complete, but part of that process is not to allow Alice to delete nodes.

That said, we should probably query the user's permissions and mention that they can't do what they are trying to do. 

Or, we let `down` run in the foreground and report back on its success.

mplsgrant commented:
> will need to dig into the commander image more, but kinda weird that the commander image needs to create "secrets"?

@josibake The other move right now might be to get rid of Alice and her hobbled permissions. She makes the `down` command not work, and we get this secrets error in her account. Do we really need a user with her permissions?

mplsgrant committed:
testing: remove mention of minikube

mplsgrant commented:
> test fails locally bc im using docker desktop:

@pinheadmz Thank you. Fixed here: https://github.com/bitcoin-dev-project/warnet/pull/616/commits/c638a1abbb1ca5b9fa32ae0c4fc37ff8905c1977 "testing: remove mention of minikube"

josibake commented:
> Do we really need a user with her permissions?

Permissions are configuration based (mostly), so I'd expect either warnet to tell me I'm not allowed to do something or complete the action successfully, based on my permissions. FWIW, I switched to Bob and also tried to run a scenario. Didn't get the permissions error but the container immediately went to a succeeded status, which seems wrong?

mplsgrant committed:
tesing: use a temporary directory

mplsgrant committed:
admin: update create_kubeconfig description

mplsgrant committed:
admin: spelling nit

mplsgrant committed:
DRY out the namespace check

Replace the namespace check with this function: `get_default_namespace_or(namespace)`

mplsgrant committed:
testing: finish replacing minikube logic

mplsgrant committed:
ruff get_default_namespace_or

mplsgrant committed:
testing: bring service account checking "in house"

mplsgrant committed:
service_accounts: remove func from `admin` section

mplsgrant committed:
admin: make kubeconfig a dict

And use yaml.dump

mplsgrant committed:
constants: use labels to select pods for `log`

mplsgrant committed:
rename to `get_namespaces_by_type`

`get_namespaces_by_prefix` is now renamed to `get_namespaces_by_type` in anticipation of using
labels in the future.

mplsgrant committed:
k8s: add `can_delete_pods` function

We use this to check if the current user can delete pods.

mplsgrant committed:
control: update `down` with `can_delete_pods`

We want to back out early if the user is not able to bring down the network

mplsgrant commented:
> Also, when running `warnet down`, it doesn't actually delete any resources (expected, per the permissions we gave Alice), but it tells the user that it did do the teardown successfully (lies!):

@josibake Fixed here by checking user permissions: https://github.com/bitcoin-dev-project/warnet/pull/616/commits/8ad00cd2f42f41667ee4d7c92588d57727d04946 and https://github.com/bitcoin-dev-project/warnet/pull/616/commits/4c62c79294e0dad5cc1ac5386b279a1a92322f50
"k8s: add can_delete_pods function" and "Using team namespaces: Kubeconfig Step Thru"

m3dwards commented:
Test now passes for me on Docker Desktop. Was having multiple local k8s that people use for dev useful here to drive out a problem??? haha.

pinheadmz commented on test/namespace_admin_test.py:
```diff
@@ -0,0 +1,127 @@
+        self.log.info("Authenticating and becoming bob...")
+        assert get_kubeconfig_value("{.current-context}") == self.initial_context
+        self.log.info(self.warnet("auth kubeconfigs/bob-wargames-red-team-kubeconfig"))
+        assert get_kubeconfig_value("{.current-context}") == "bob-wargames-red-team"
```
just so I'm clear: a context is something like `docker-desktop` but also its `username + namespace` ?

mplsgrant replied:
My understanding is...

context is a cluster ("minikube", "digital-ocean-string", etc) + username ("john") + optionally a namespace ("default", "wargames-red-team", etc)

Also, we store contexts in the kubeconfig file.

m3dwards commented on docs/admin.md:
````diff
@@ -0,0 +1,70 @@
+
+```shell
+$ cd namespaces
+$ warnet deploy two_namespaces_two_users
````
Perhaps:

```suggestion
$ warnet deploy namespaces/two_namespaces_two_users
```

mplsgrant replied:
Thank you. Fixed: https://github.com/bitcoin-dev-project/warnet/pull/616/commits/4460f20a6e927fcb9e49bbf573f15dafd5ed50e4 "admin.md: update namespace deploy command"

mplsgrant committed:
admin.md: update namespace deploy command

mplsgrant committed:
bitcoin: make ruff happy

mplsgrant commented:
rebased on main

m3dwards commented:
I think the test should either cleanup it's modifications to your kubeconfig or use a temporary kubeconfig.

m3dwards commented:
Miner scenario runs for me when I'm authenticated as one of the users.

m3dwards commented:
I've not poured over the code, more looked at this from a functional perspective. It's working for me, left a few comments.

mplsgrant committed:
k8s: add open/write kubeconfig fn; add K8sError

The K8sError allows us to easily handle one single error that crops up from functions that life in
k8s.py. Right now I just have it implemented in the open/write kubeconfig functions.

mplsgrant committed:
testing: specify `warnettest` in named items

pinheadmz commented:
Using `warnet admin init` to update my battle repo:
- i wish it ONLY created the `namespaces/` directory. i already have `networks/` and i dont need another copy of `6_node...`. It also doesn't copy the default `scenarios/` so its incomplete.
- i wish it did the inquirer thing like `warnet create` does. otherwise i just get plopped an example namespace, and i have to reverse-engineer that example to make what i need

m3dwards commented:
> * i wish it ONLY created the `namespaces/` directory

aren't you likely to create a network so having a reference there is handy? I mean we could copy namespaces in too, the point of the init is to copy some examples?

As for not copying scenarios, I can understand why it doesn't. It wouldn't be expected that an admin would be running scenarios?

m3dwards commented:
I think this PR should be updated to remove `--create-namespace` from the helm commands. This prevents a user from deploying their own network.

pinheadmz commented:
> aren't you likely to create a network so having a reference there is handy? I mean we could copy namespaces in too, the point of the init is to copy some examples?

i had already `warnet create`  a few weeks ago and i am fiddling with my networks. today i decide to add namespaces and i end up with this:

```
/
  networks/
  scenarios/
  admin/
    networks/
    namespaces/
```

pinheadmz commented:
getting this now, but i have moved some files around ...?


```
(.venv) --> warnet deploy namespaces/armies/
Deploying namespace: wargames-red
Release "wargames-red" does not exist. Installing it now.
NAME: wargames-red
LAST DEPLOYED: Fri Oct  4 13:26:51 2024
NAMESPACE: default
STATUS: deployed
REVISION: 1
TEST SUITE: None
Traceback (most recent call last):
  File "/Users/matthewzipkin/Desktop/work/warnet/.venv/bin/warnet", line 8, in <module>
    sys.exit(cli())
             ^^^^^
  File "/Users/matthewzipkin/Desktop/work/warnet/.venv/lib/python3.12/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/matthewzipkin/Desktop/work/warnet/.venv/lib/python3.12/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/Users/matthewzipkin/Desktop/work/warnet/.venv/lib/python3.12/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/matthewzipkin/Desktop/work/warnet/.venv/lib/python3.12/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/matthewzipkin/Desktop/work/warnet/.venv/lib/python3.12/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/matthewzipkin/Desktop/work/warnet/src/warnet/deploy.py", line 62, in deploy
    _deploy(directory, debug, namespace, to_all_users)
  File "/Users/matthewzipkin/Desktop/work/warnet/src/warnet/deploy.py", line 83, in _deploy
    deploy_namespaces(directory)
  File "/Users/matthewzipkin/Desktop/work/warnet/src/warnet/deploy.py", line 298, in deploy_namespaces
    temp_override_file_path.unlink()
  File "/opt/homebrew/Cellar/python@3.12/3.12.5/Frameworks/Python.framework/Versions/3.12/lib/python3.12/pathlib.py", line 1342, in unlink
    os.unlink(self)
PermissionError: [Errno 1] Operation not permitted: '.'
```

pinheadmz commented:
sorry for spamming the PR comments but I think this is a necessary fix that should be included:

```diff
diff --git a/src/warnet/deploy.py b/src/warnet/deploy.py
index 63c888e3..f32156c4 100644
--- a/src/warnet/deploy.py
+++ b/src/warnet/deploy.py
@@ -273,7 +273,7 @@ def deploy_namespaces(directory: Path):
     for namespace in namespaces_file["namespaces"]:
         click.echo(f"Deploying namespace: {namespace.get('name')}")
         try:
-            temp_override_file_path = Path()
+            temp_override_file_path = ""
             namespace_name = namespace.get("name")
             namespace_config_override = {k: v for k, v in namespace.items() if k != "name"}
 
@@ -294,7 +294,7 @@ def deploy_namespaces(directory: Path):
             click.echo(f"Error: {e}")
             return
         finally:
-            if temp_override_file_path.exists():
+            if temp_override_file_path:
                 temp_override_file_path.unlink()
 
 
```

pinheadmz commented on docs/admin.md:
````diff
@@ -0,0 +1,69 @@
+
+
+```shell
+$ warnet admin deploy networks/mynet --to-all-users
````
I think this is just `deploy` not `admin deploy` ?

m3dwards commented:
> i had already `warnet create`

This is a good point. If you want networks, you can just call warnet init.

mplsgrant committed:
testing: update test to include cleanup

mplsgrant committed:
auth: update auth func to avoid flattening

mplsgrant committed:
k8s: get raw config values from kubectl

mplsgrant committed:
k8s: get cluster from current context

mplsgrant committed:
constants: remove --create-namespace

mplsgrant committed:
admin: get raw cluster for auth file

pinheadmz commented:
Another thing that would make the battle easier to set up (for me) is if the namespaces list also included a "battlefield" for the admin to use. Then I can confidently hardcode tank names like `miner.battlefield` etc my flow is looking like this:

1. generate network files: 120-tank battlefield, 3-node armada
2. generate namespaces: 1 for each team, 1 battlefield for admin
3. generate config files for each namespace
4. `warnet auth battlefield && warnet deploy networks/battlefield` as admin
5. then players will `warnet auth wargames-red-team` and then can run scenarios

pinheadmz commented:
I think the initial warning here should list literally ALL namespaces that will be down'ed

```
(.venv) --> warnet down
Service account can delete pods in namespace 'default'.
Preparing to bring down the running Warnet...
The listed namespaces will be affected:
  default
[?] Do you want to bring down the running Warnet? (y/N): y

Bringing down the warnet...
Initiated uninstall for: wargames-white in namespace default
Initiated uninstall for: wargames-violet in namespace default
Initiated uninstall for: wargames-black in namespace default
Initiated uninstall for: wargames-battlefield in namespace default
Initiated uninstall for: wargames-yellow in namespace default
Initiated uninstall for: wargames-orange in namespace default
Initiated uninstall for: wargames-blue in namespace default
Initiated uninstall for: wargames-green in namespace default
Initiated uninstall for: wargames-brown in namespace default
Initiated uninstall for: wargames-red in namespace default
Initiated uninstall for: wargames-grey in namespace default
Teardown process initiated for all components.
Note: Some processes may continue in the background.
Warnet teardown process completed.

```

pinheadmz commented:
Ok i see, so my official request is that the hard coded `namespaces/` chart we template from always includes an admin service account that can admin ALL namespaces as well as warnet-logging


> Another thing that would make the battle easier to set up (for me) is if the namespaces list also included a "battlefield" for the admin to use. Then I can confidently hardcode tank names like `miner.battlefield` etc my flow is looking like this:
> 
>     1. generate network files: 120-tank battlefield, 3-node armada
> 
>     2. generate namespaces: 1 for each team, 1 battlefield for admin
> 
>     3. generate config files for each namespace
> 
>     4. `warnet auth battlefield && warnet deploy networks/battlefield` as admin
> 
>     5. then players will `warnet auth wargames-red-team` and then can run scenarios

mplsgrant commented:
> I think the test should either cleanup it's modifications to your kubeconfig or use a temporary kubeconfig.

@m3dwards I have created a cleanup solution and pushed it.

mplsgrant commented:
> I think this PR should be updated to remove `--create-namespace` from the helm commands. This prevents a user from deploying their own network.

@m3dwards updated. Thank you!

pinheadmz commented:
`warnet messages` doesn't seem to work across namespaces. Auth'ed as red team, checking messages from miner in battlefield: (also tried `miner.default`)

```
(.venv) --> warnet bitcoin messages armada-0 miner
Error fetching messages between nodes armada-0 and miner: Error from server (NotFound): pods "miner" not found
```

mplsgrant committed:
constants: does graph_test.py need --create-namespace?

mplsgrant commented:
> `warnet messages` doesn't seem to work across namespaces. Auth'ed as red team, checking messages from miner in battlefield: (also tried `miner.default`)

You would need to add the --namespace_a and --namespace_b flags I believe

pinheadmz commented:
would be super cool if it were just `warnet bitcoin messsges tank-0000-red
miner.default` (using the dot notation for namespace)

On Fri, Oct 4, 2024 at 3:06 PM mplsgrant ***@***.***> wrote:

> warnet messages doesn't seem to work across namespaces. Auth'ed as red
> team, checking messages from miner in battlefield: (also tried
> miner.default)
>
> You would need to add the --namespace_a and --namespace_b flags I believe
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin-dev-project/warnet/pull/616#issuecomment-2394393347>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAP46KG27OHC3OMXLYA5WR3ZZ3RJXAVCNFSM6AAAAABOXK5Q6GVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDGOJUGM4TGMZUG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

mplsgrant committed:
bitcoin: update `message` to take tank-a.namespace

mplsgrant commented:
> would be super cool if it were just `warnet bitcoin messsges tank-0000-red miner.default` (using the dot notation for namespace)

@pinheadmz Good idea. That now works, except you an leave off "default" as it will automatically tack that on if it doesn't see a namespace.

m3dwards commented:
Anything outstanding?

pinheadmz commented:
Did this get addressed ?

https://github.com/bitcoin-dev-project/warnet/pull/616#issuecomment-2394264911

mplsgrant committed:
deploy: fix override path

mplsgrant committed:
`admin.md`: reword deploy documentation

mplsgrant commented:
> Did this get addressed ?
> 
> [#616 (comment)](https://github.com/bitcoin-dev-project/warnet/pull/616#issuecomment-2394264911)

@pinheadmz I have now pushed a commit to address that.

mplsgrant commented:
@m3dwards 
Another open item is how we handle `--create-namespace`. It currently remains in the code base.

mplsgrant commented:
Also, note that I had updated the `auth` function and `create_kubeconfigs` in the latest commits: https://github.com/bitcoin-dev-project/warnet/pull/616/commits/0d45104f2fae6bd2a39633e1962ef5d3090badef https://github.com/bitcoin-dev-project/warnet/pull/616/commits/e00ee5eaf6a3259dbb8e8dcae2e091209ed98b1e https://github.com/bitcoin-dev-project/warnet/pull/616/commits/4f1245a4e7e14e856617282a28d6a89f439ed77b https://github.com/bitcoin-dev-project/warnet/pull/616/commits/4d77dff7142a777cd061fd02b974711a9cefe5e4

I also had updated the `messages` function to take "tank.namespace" parameters: https://github.com/bitcoin-dev-project/warnet/pull/616/commits/191da0b7c6f078110d0e92da5fd4201ae5e7f493

m3dwards committed:
removing --create-namespace from constants

m3dwards committed:
removing --create-namespace from constants

mplsgrant committed:
gitignore: add kubeconfigs dir

mplsgrant committed:
k8s: use mv to prevent scenario getting cut off

We use this following to check for the scenario file:

          while [ ! -f /shared/archive.pyz ]; do
            echo "Waiting for /shared/archive.pyz to exist..."
            sleep 2
          done

I suspect this may cut off longer-running copy commands. Using mv we make sure all the data is
copied over, and then we `mv` the file into place.

mplsgrant committed:
k8s: add `sync` to make sure the data is written

mplsgrant commented:
rebased on main
